File: dynamic-programming/cpp/interleaving.cpp
/**
    * @author 	Luca Cortinovis	luca1cortinovis@gmail.com
    * @since	28/10/2023
    *
    * You can find me here:
    * GitHub	Luca Cortinovis (NeptuneLuke)
*/

#include <iostream>
#include <string>
#include <vector>


using
std::string,
std::vector,
std::cout,
std::cin;


bool is_interleaving(string s1, string s2, string interleaving);


int main(void) {

    string s1, s2, s3;
    cout << "Insert the first string \n";
    getline(cin, s1);

    cout << "Insert the second string \n";
    getline(cin, s2);

    cout << "Insert the interleaving string \n";
    getline(cin, s3);

    bool result = is_interleaving(s1, s2, s3);
    cout << "Is " << s3 << " interleaving? \n";
    cout << std::boolalpha << result;


    return 0;
}

bool is_interleaving(string s1, string s2, string interleaving) {

    int m = s1.size();
    int n = s2.size();
    vector< vector<bool> > interleaving_table (m + 1, vector<bool>(n + 1, false) );


    if (m + n != interleaving.size()) {
        return false;
    }

    for (int i = 0; i <= m; ++i) {

        for (int j = 0; j <= n; ++j) {

            // base case i==0 and j==0
            if (i == 0 && j == 0) {
                interleaving_table[i][j] = true;
            }
            // S1 is empty
            else if (i == 0) {
                if (s2[j-1] == interleaving[j-1])
                    interleaving_table[i][j] = interleaving_table[i][j-1];
            }
            // S2 is empty
            else if (j == 0) {
                if (s1[i-1] == interleaving[i-1])
                    interleaving_table[i][j] = interleaving_table[i-1][j];
            }
            // S1[i] == interleaving[i] and S2[i] != interleaving[i]
            else if (s1[i-1] == interleaving[i+j-1] && 
                     s2[j-1] != interleaving[i+j-1]) {
                interleaving_table[i][j] = interleaving_table[i-1][j];
            }
            // S1[i] != interleaving[i] and S2[i] == interleaving[i]
            else if (s1[i-1] != interleaving[i+j-1] &&
                s2[j-1] == interleaving[i+j-1]) {
                interleaving_table[i][j] = interleaving_table[i][j-1];
            }
            // S1[i] == S2[i] == interleaving[i]
            else if (s1[i-1] == interleaving[i+j-1] && 
                     s2[j-1] == interleaving[i+j-1]) {
                interleaving_table[i][j] = (interleaving_table[i-1][j] || interleaving_table[i][j-1]);
            }
        }
    }

    return interleaving_table[m][n];
}

File: dynamic-programming/java/Interleaving.java
/**
* @author 	Luca Cortinovis	luca1cortinovis@gmail.com
* @version	1.0
* @since	9/11/2022
*
* You can find me here:
* GitHub	Luca Cortinovis(NeptuneLuke)
*/


import java.util.Scanner;

public class Interleaving {

	
	public static void main(String args[]) {
		
		Scanner keyboard = new Scanner(System.in);
		
		System.out.println("Insert first sequence: ");
		String X = keyboard.next();
		
		System.out.println("Insert second sequence: ");
		String Y = keyboard.next();
	
		System.out.println("Insert interleaving sequence: ");
		String W = keyboard.next();
		
		System.out.println("Is " + W + " interleaving of " + X + " " + Y + " --> " + isInterleave(X,Y,W));
		
	}
	
	
	/* Iterative algorithm that fills matrix S with the solutions of all subproblems (Xi,Yj,Wi+j)
	 * Time complexity: O(m*n) 
	 */
	public static boolean isInterleave(String X, String Y, String W) {
		
		int m = X.length();
		int n = Y.length();
		
		boolean S[][] = new boolean[m+1][n+1];
		
		// by definition if W is interleaving of X,Y
		// W.length() == X.length() + Y.length()
		if ((m+n) != W.length())
			return false;
		
		// W.charAt(i+j-1) current char of W
		// X.charAt(i-1) current char of X
		// Y.charAt(j-1) current char of Y
		for(int i=0; i<=m; i++) {
			for(int j=0; j<=n; j++) {
		
				if(i == 0 && j == 0)		// base case
					S[i][j] = true;
				
				else if(i == 0 && j > 0) {	// base case
					
					if(W.charAt(i+j-1) == Y.charAt(j-1))
						S[i][j] = S[i][j-1];
					else
						S[i][j] = false;
				}
				else if(j == 0 && i > 0) {	// base case
					
					if(W.charAt(i+j-1) == X.charAt(i-1))
						S[i][j] = S[i-1][j];
					else
						S[i][j] = false;
				}
				
				// "recursive step" (i,j) > 0
				else {
					
					if(W.charAt(i+j-1) != X.charAt(i-1) && W.charAt(i+j-1) != Y.charAt(j-1))		// W(i+j) != X(i) AND W(i+j) != Y(j)
						S[i][j] = false;
					else if(W.charAt(i+j-1) == X.charAt(i-1) && W.charAt(i+j-1) != Y.charAt(j-1))	// W(i+j) == X(i) AND W(i+j) != Y(j)
						S[i][j] = S[i-1][j];
					else if(W.charAt(i+j-1) != X.charAt(i-1) && W.charAt(i+j-1) == Y.charAt(j-1))	// W(i+j) != X(i) AND W(i+j) == Y(j)
						S[i][j] = S[i][j-1];
					else if(W.charAt(i+j-1) == X.charAt(i-1) && W.charAt(i+j-1) == Y.charAt(j-1))	// W(i+j) == X(i) AND W(i+j) == Y(j)
						S[i][j] = S[i-1][j] || S[i][j-1];
				}
					
			}
		}
		
		return S[m][n];
	}
	
	
}
